Wrappers/Python/cil/optimisation/algorithms/SPDHG.py

```diff
@@ -42,202 +46,460 @@ class SPDHG(Algorithm):
     operator : BlockOperator
         BlockOperator must contain Linear Operators
     tau : positive float, optional, default=None
-        Step size parameter for Primal problem
+        Step size parameter for primal problem. If `None` see note.
     sigma : list of positive float, optional, default=None
-        List of Step size parameters for Dual problem
-    initial : DataContainer, optional, default=None
+        List of Step size parameters for dual problem. If `None` see note.
+    initial : DataContainer, optional, the default value is a zero DataContainer in the range of the `operator`.
         Initial point for the SPDHG algorithm
-    prob : list of floats, optional, default=None
-        List of probabilities. If None each subset will have probability = 1/number of subsets
-    gamma : float
-        parameter controlling the trade-off between the primal and dual step sizes
+    gamma : float, optional
+            Parameter controlling the trade-off between the primal and dual step sizes
+    sampler: optional, an instance of a `cil.optimisation.utilities.Sampler` class or another class with the function __next__(self) implemented outputting an integer from {1,...,len(operator)}. 
+            Method of selecting the next index for the SPDHG update. If None, a sampler will be created for random sampling with replacement and each index will have `probability = 1/len(operator)`
+    prob_weights: optional, list of floats of length `num_indices` that sum to 1. Defaults to `[1/len(operator)]*len(operator)`
+            Consider that the sampler is called a large number of times this argument holds the expected number of times each index would be called,  normalised to 1. Note that this should not be passed if the provided sampler has it as an attribute: if the sampler has a `prob_weight` attribute it will take precedence on this parameter. 
```

Comment:
the input type needs to be concise. `list of floats, optional` with the description expanding.

Beyond that, it's not clear why we need this and where it's used. ISn't this what sampler now controls? So either it's a docstring or implementation problem. Should it be moved to `kwargs`?

Reply:
There was a design decision made that the sampler doesn't have to have `prob_weights` as in other stochastic algorithms they are not essential, just for reporting and plotting. However, this algorithm requires them to set `sigma` and `tau`. As the sampler might not have that argument, instead it can be passed seperately to the algorithm. But maybe we chat about it being a kwarg...